File: maximum_value_of_k_coins_from_piles.cpp
class Solution {
public:
    int maxValueOfCoins(vector<vector<int>>& piles, int k) {
        std::vector<int> prev(k + 1, 0);

        for (const auto& pile : piles) {
            std::vector<int> sumList(1, 0);
            int total = 0;
            for (int coin : pile) {
                total += coin;
                sumList.push_back(total);
            }

            std::vector<int> curMax(k + 1, 0);

            for (int n = 1; n <= k; n++) {
                int calcRange = std::min(n + 1, static_cast<int>(sumList.size()));
                for (int pos = 0; pos < calcRange; pos++) {
                    curMax[n] = std::max(curMax[n], sumList[pos] + prev[n - pos]);
                }
            }

            prev = curMax;
        }

        return prev[k];        
    }
};

File: maximum_value_of_k_coins_from_piles.py
class Solution:
    def maxValueOfCoins(self, piles: List[List[int]], k: int) -> int:

        prev = [0] * (k + 1)

        for pile in piles:
            sum_list = [0]
            total = 0
            for coin in pile:
                total += coin
                sum_list.append(total)
        
            cur_max = [0] * (k + 1)

            for n in range(1, k + 1):
                calc_range = min(n + 1, len(sum_list))
                cur_max[n] = max(sum_list[pos] + prev[n - pos] for pos in range(calc_range))
            
            prev = cur_max
        
        return prev[k]
